### DISCUSSION TITLE
如何精准控制http client的超时

### DISCUSSION THREAD

**Discussion by fengqi** [swoole/swoole-src#5065] [Q&A]
使用[协程客户端](https://wiki.swoole.com/#/coroutine_client/http_client) ，超时设置为1.0，实际执行时间能达到1300ms。

我看[文档](https://wiki.swoole.com/#/coroutine_client/init?id=%e8%b6%85%e6%97%b6%e8%a7%84%e5%88%99)描述：这个是总超时，包括连接、发送、接收所有超时，那这多出来的300ms时间花在哪里了呢，DNS解析？如果我想要更精准的控制超时，还需要设置哪些参数呢

**Comment by NathanFreeman** [comment]
这300ms应该是其他业务逻辑的执行时间吧，1秒内没完成连接，然后清理定时器，将结果返回回去。

**Reply by fengqi** [reply]
我测试过直接新建一个client，没有任何业务逻辑现象也是一样的：
```
<?php

go(function () {
    $s = microtime(true);
    $client = new \Swoole\Coroutine\Http\Client('xxx.com', 443, true);
    $client->set(['timeout' => 1.0]);
    $client->setMethod('GET');
    $client->execute('/sleep.php'); // 目标接口sleep 1s模拟超时
    var_dump($client->getBody(), $client->getStatusCode(), microtime(true) - $s); // 拿不到返回，code=-2，时间1300左右
});
```

**Comment by NathanFreeman** [comment]
你这个`$s`是在最开始的地方初始化的，请求之前的准备工作的时间和超时之后的后续工作时间全部计算在里面了。
超时之后不是直接中断，中断之后swoole内部也要做一些后续工作的，包括将定时器删除，返回结果给上层应用。

**Reply by fengqi** [reply]
$s放到execute前一行最终结果也是一样的，初始化应该也没做啥，我从0.1-1.0测下来基本上都会浪费300ms左右的时间。
swoole内部到底做了啥我应该干涉不了，我比较想知道如果想精准控制超时应该咋做，把所有设置的超时打个0.7折？感觉这不太优雅

**Comment by fengqi** [comment]
用curl基本上误差只有几ms，还挺精准的

**Comment by twose** [comment]
cURL 是同步阻塞，通过外部定时打断的，所以精准。
你也可以再起一个协程，在一定时间后打断请求的协程。

**Reply by fengqi** [reply]
现在是这样做的，感觉不够优雅😂

**Reply by twose** [reply]
https://heapdump.cn/article/4934972

随便搜了一篇文章，不出意外的，golang 也是这么实现的，而且更重。
